kubernetes-exec-helper/src/Kubernetes/WSClient.hs

```diff
@@ -0,0 +1,204 @@
+
+import Control.Concurrent(ThreadId)
+import Control.Concurrent.Async(waitAny, async, Async)
+import Control.Concurrent.STM
+import Control.Exception.Safe
+import Control.Monad (forever)
+import Control.Monad.Reader
+import Data.Maybe
+import Data.Monoid ((<>))
+import Data.Text (Text)
+import Kubernetes.K8SChannel
+import Kubernetes.KubeConfig
+import Network.Socket(withSocketsDo)
+import qualified Data.Text as T
+import qualified Data.Text.IO as T
+import qualified Network.WebSockets as WS
+import System.Timeout (timeout) 
+import System.IO (hSetBuffering, BufferMode(..), stdin)
+
+-- | Run the web socket client. 
+runClient :: String -- ^ Host  
+            -> Int  -- ^ Port 
+            -> String -- ^ Path
+            -> Maybe TimeoutInterval -- ^ Channel timeout.
+            -> IO (CreateWSClient Text)
+runClient domain port route = \timeout' -> 
+  withSocketsDo $ WS.runClient domain port route $ 
+    (\c -> bracket 
+          (return c)
+          (\conn ->WS.sendClose conn ("Closing session" :: T.Text))
+          (\conn -> k8sClient conn timeout'))
+
+-- | Socket IO handler.
+k8sClient :: WS.Connection -> Maybe TimeoutInterval -> IO (CreateWSClient Text)
+k8sClient conn interval = do
+    cW <- atomically newTChan :: IO (TChan Text) -- Writer channel
+    c0 <- atomically newTChan
+    c1 <- atomically newTChan 
+    c2 <- atomically newTChan 
+    c3 <- atomically newTChan 
+    c4 <- atomically newTChan
+    let channels = zip allChannels [c0, c1, c2, c3, c4]
+    rcv <- timedThread worker channels interval
+    sender <- async $ forever $ do 
+        nextMessage <- atomically . readTChan $ cW 
+        WS.sendTextData conn nextMessage
+    return . CreateWSClient $ (catMaybes[Just sender, rcv], cW, channels)
+    where
+      worker channels = async $ forever $ do 
+            msg <- (WS.receiveData conn) `catch` (\e@(SomeException _) -> return $ T.pack . show $ e)
+            publishMessage channels $ T.splitAt 1 msg
+      timedThread aWorker channels timeoutInterval =
+        case timeoutInterval of  
+            Nothing ->  Just <$> aWorker channels
+            Just m -> timeout m $ aWorker channels
+
+-- | Publish messages from the reader into the channel.
+publishMessage :: [(ChannelId, TChan Text)] -> (Text, Text) -> IO ()
+publishMessage channels (channel, message) = do 
+  let chanId = readChannel channel
+  case chanId of
+    Nothing -> throwIO $ InvalidChannel channel
+    Just aChan -> 
+      atomically $
+        writeTChan (snd $ getChannelIdSTM aChan channels) message
+
+{- | 
+  Query a channelId from a list of Channels.
+-}
+getTChanSTM :: ChannelId -> [(ChannelId, TChan Text)] -> TChan Text 
+getTChanSTM a b = snd $ getChannelIdSTM a b 
+getChannelIdSTM :: ChannelId -> [(ChannelId, TChan Text)] -> (ChannelId, TChan Text)
+getChannelIdSTM aChannelId channels = 
+  head $ Prelude.filter(\(x, _) -> x == aChannelId) channels
+
+readChannelIdSTM :: ChannelId -> [(ChannelId, TChan Text)] -> STM Text 
```

Comment:
What's the purpose of these read/write functions? I don't see them used anywhere.